### PR TITLE
One press, one shot

### LINKED ISSUE #3: Fix rate of fire
Make the player fire on shot per spacebar.

### PR DESCRIPTION
**Pull request by Matt-Warnock**
Rather than a rate of fire, I have built into the Keyboarder a method to shot only once when the key is pressed.

### CODE REVIEW

File: keyboarder.js [hunk 1]
```diff
@@ -2,16 +2,29 @@
...
-    };
-    this.isDown = function(keyCode) {
-      return keyState[keyCode] === true;
+    window.onkeyup = (e) => {
+      this.keyState[e.keyCode] = false;
     };
   }
+
+  isDown(keyCode) {
+    return this.keyState[keyCode] === true;
```
Comment (octopusinvitro) [inline keyboarder.js]: since `this.keyState` only stores booleans, you can just return the value:

```js
return this.keyState[keyCode] ;
```

File: keyboarder.js [hunk 2]
```diff
@@ -2,16 +2,29 @@
...
+  keyTap(keyCode) {
+    if (this.keyState[keyCode] === false) {
+      this.isKeyUp = true;
+    }
+    if (!this.isKeyUp) return;
+    else if (this.keyState[keyCode] === true) {
+      this.isKeyUp = false;
+      return true;
+    }
+  }
```
Comment (octopusinvitro) [inline keyboarder.js:24]: * If `this.isKeyUp` is `false`, you are returning null, this is surprising for the developer as this method can also return `true`. It's better to make it return a boolean always, `false` in this case

    (Regardless of if the languages allows you to return different types from the same function, it is better for the sake of readability and brain cycles to return the same type.)

* Again, you don't need to do `this.keyState[keyCode] === true`, `this.keyState[keyCode]` already returns a boolean, you can just do `if(this.keyState[keyCode])`.

So what is this method trying to do?

* if `this.keyState[keyCode]` is false, `this.isKeyUp` is true, and viceversa. That is just one line of code:

```js
  this.isKeyUp = !this.keyState[keyCode];
```

we did this in the address book already.

* The other thing this method is doing is returning the opposite value of `this.isKeyUp`. So if it is true, it returns false and viceversa. Who has the opposite value of `this.isKeyUp`? `this.keyState[keyCode]`. So you can just return that. 

* Also, you only need this method to work when the user presses the spacebar, so no need you pass the `keyCode` at all. Then you can rename it with a less generic name, like `isShooting()`.
 
In other words, this method can be simplified as:

```js
  isShooting() {
    this.isKeyUp = !this.keyState[this.KEYS.SPACE];
    return this.keyState[this.KEYS.SPACE];
  }
```

As you can see, you are setting `this.isKeyUp`, but you are never using it anywhere. Hence, you don't really need that variable, and can simplify this method even more:

```js
  isShooting() {
    return this.keyState[this.KEYS.SPACE];
  }
```

File: player.js [hunk 3]
```diff
@@ -5,8 +5,9 @@ class Player {
     this.margin = 2;
     this.size = {x: 15, y: 15};
     this.center = {x: gameSize.x / 2, y: gameSize.y - this.size.x};
-    this.keyboarder = new Keyboarder();
+    this.keyboarder = new Keyboarder(game);
```
Comment (octopusinvitro) [inline player.js]: The Keyboarder does not take a game as argument in its constructor.

**Review by octopusinvitro — COMMENTED** [review]
You didn't need `this.isKeyUp` after all!

File: player.js (hunk 3 above)
Comment (Matt-Warnock) [reply player.js]: whoops, forgot to get rid of that

File: keyboarder.js (hunk 2 above)
Comment (Matt-Warnock) [reply keyboarder.js:24]: returning null isn't good. I've changed that. Also have refactored the "this.keyState[keyCode] === true" as I understand that is unnecessary.

as I understand it, this.isKeyUp is needed to log the keys state in other instances the program runs. I've made some videos I'll send to you at the end of the day as I'm sure you're busy.

Comment (Matt-Warnock) [reply keyboarder.js:29]: So what is this method trying to do?

we can talk about this question later on as well if you like.

Comment (octopusinvitro) [reply keyboarder.js:24]: can you link me to the other places in the code where `isKeyUp` is used? no one is calling that variable, maybe you didn't push? :thinking:

Comment (octopusinvitro) [reply keyboarder.js:24]: I see what is the problem.
You are making things more complicated that they need to be.
All you need to do is to reset the spacebar after you first detect that it's down, by setting `this.keyState[keyCode]` to `false`.

Like this:

In the player class:

```js
    if (this.keyboarder.isDown(this.keyboarder.KEYS.SPACE)) {
      this.keyboarder.restore(this.keyboarder.KEYS.SPACE);
      let bullet = new Bullet ({x: this.center.x, y: this.center.y - this.size.x / 2}, {x: 0, y: -7});
      this.game.addBody(bullet);
    }
```

in the keyboarder class:

```js
  restore(keyCode) {
    this.keyState[keyCode] = false;
  }
```

and you can ditch that `this.isKeyUp` variable. There is no need for it.

File: player.js [hunk 4]
```diff
@@ -1,10 +1,9 @@
 class Player {
   constructor(game, gameSize) {
     this.game = game;
-    this.gameLeftSide = 0;
     this.margin = 2;
     this.size = {x: 15, y: 15};
-    this.center = {x: gameSize.x / 2, y: gameSize.y - this.size.x};
+    this.center = {x: gameSize.x / this.margin, y: gameSize.y - this.size.x};
```
Comment (octopusinvitro) [inline player.js]: this is not the margin here, you are halving the horizontal game size.
If the margin changes, you don't want to be suddenly dividing it by 42!

File: player.js [hunk 5]
```diff
@@ -13,13 +12,13 @@ class Player {
...
-    if (this.center.x - this.size.x / this.margin <= this.gameLeftSide) {
+    if (this.center.x - this.size.x / this.margin <= this.game.gameLeftSide) {
       this.center.x += 2;
     }else if (this.center.x + this.size.x / this.margin >= this.game.gameSize.x) {
       this.center.x -= 2;
     }
-    if (this.keyboarder.isDown(this.keyboarder.KEYS.SPACE)) {
-      let bullet = new Bullet ({x: this.center.x, y: this.center.y - this.size.x / 2}, {x: 0, y: -7});
+    if (this.keyboarder.shootOnce(this.keyboarder.KEYS.SPACE)) {
+      let bullet = new Bullet ({x: this.center.x, y: this.center.y - this.size.x / this.margin}, {x: 0, y: -7});
```
Comment (octopusinvitro) [inline player.js]: this is half the horizontal size!
don't divide by the margin, it just happens to be 2px now, but it could change

**Review by octopusinvitro — CHANGES_REQUESTED** [review]
The new implementation is again way too complicated. You are not adding one extra variable, but two! And have extra if/else complicated logic in the method. When you find yourself adding this complex logic, stop, take a step back, and rethink.

Remove those two variables and the new methods and go back to [the suggestion I gave you](https://github.com/Matt-Warnock/space_invaders/pull/12#discussion_r430704704), which is the simplest way to stop the player from shooting after the first shoot.

Then, to stop the player from shooting when the key is constantly pressed, all you have to add is this to the keydown listener:
```js
    window.onkeydown = (e) => {
      this.keyState[e.keyCode] = true;

      if (e.repeat && e.keyCode === this.KEYS.SPACE) {
        this.keyState[e.keyCode] = false;
      }
    };
```

File: player.js (hunk 5 above)
Comment (Matt-Warnock) [reply player.js]: I think I did that on request on a previous pull request.
I have now changed it back to 2.

**Review by octopusinvitro — APPROVED** [review]
:+1: 
remember to rebase this branch against master before merging:

```
git checkout rate_of_fire
git rebase master
```

and fix any conflicts that may appear :slightly_smiling_face: 

PS. Regarding the margin, use `2` explicitly when halving dimensions, and `this.margin` when setting the position of the player. Like this:

```js
    if (this.center.x - this.size.x / 2 <= this.gameLeftSide) {
      this.center.x = this.margin;
    } else if (this.center.x + this.size.x / 2 >= this.game.gameSize.x) {
      this.center.x = this.game.gameSize.x - this.margin;
    }
```
